README
NAME
    Net::Minecraft::Login - Basic implementation of the Minecraft Login
    Protocol.

VERSION
    version 0.001001

DESCRIPTION
    This is a basic implementation of the Minecraft Login protocol as
    described at <http://wiki.vg/Session#Login>

      use Net::Minecraft::Login;

      my $ua = Net::Minecraft::Login->new();

      my $result = $ua->login(
        user => 'Bob',
        password => 'secret',
      );
      if( not $result->is_success ){
        die $result;
      }
      say "Login for user " . $result->user . " succeeded";

    Note, it presently does no explicit session stuff, only performs the
    basic HTTP Request and returns the response as an object.

CONSTRUCTOR ARGUMENTS
    This section describes arguments that may be optionally passed to
    "->new()", but as of the time of this writing, none are explicitly
    required, and are offered only to give leverage to strange use cases (
    and tests )

      my $instance = Net::Minecraft::Login->new(
        user_agent   => ... ,
        http_headers => { ... },
        http_engine  => HTTP::Tiny->new(),
        login_server => 'https://somewhere.else.org/'
        version      => 14, # IN THE FUTURE!
      );

    Additional Constructor arguments can also be found in
    "Net::Minecraft::Role::HTTP"

  "login_server"
    HTTP Address to authenticate with.

      type  : String
      default : https://login.minecraft.net/

  "version"
    "Client" version.

      type  : String
      default : 13

    This field indicates the version of the "Launcher". Minecraft may at
    some future time produce an updated launcher, and indicate that this
    specified version is out of date.

    Mojang Minecraft Launchers will be required to download a newer version,
    and users of "Net::Minecraft::Login" will either

    a) Be required to update to a newer "Net::Minecraft::Login" that
    supports the newer version and changes that implies
    b) Assuming no "Login Protocol" Changes, only have to specify "version
    =>" either to the constructor, or as an argument to "login"

METHODS
  "login"
      signature: { user => String , password => String, version? => String }
      return   : Any( Net::Minecraft::LoginResult , Net::Minecraft::LoginFailure )

      my $result = $nmcl->login(
        user   => 'notch',
        password => 'jellybean',
      );

      if( not $result->is_success ){
        say "$result";
      } else {
        say "Logged in!";
      }

    See "::LoginFailure" and "::LoginResult"

ATTRIBUTES
  "login_server"
  "version"
PRIVATE METHODS
  "_do_request"
      signature : ( String $base_uri, Hash[ String => String ] $parameters , Hash[ String => Any ] $config )
      return    : Hash[ String => String ]

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Net-Minecraft-Login

{{$NEXT}}

0.002000 2013-12-31T05:23:58Z
 [00 Minor]
 - Component Refactoring
 - Note: This is still only supporing the legacy pre-yggdrasil api

 [Dependencies::Noteworthy]
 - Moo::Role⁺
 - Module::Build 0.4003 → 0.4203
 - Test::More 0.98 → 1.001002

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see Changes.deps{,.all,.dev} for
   details
 - build: ↑1
 - configure: ↑1
 - develop: +46 -3
 - runtime: +1
 - test: ↑1 -2

 [Documentation]
 - Added README.mkdn
 - added layout/*

 [Internals]
 - Net::Minecraft::Login has internal HTTP attributes extracted to a role
 - Net::Minecraft::Login{Failure,Result} are both Role::LoginResult

 [Modules::New]
 - Net::Minecraft::Role::HTTP
 - Net::Minecraft::Role::LoginResult

 [Packaging]
 - uses MB 0.4004 test_requires

0.001000 2013-01-06T18:34:58Z
 - First version.



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.008.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4203"
  },
  "configure_requires" => {
    "Module::Build" => "0.4203"
  },
  "dist_abstract" => "Basic implementation of the Minecraft Login Protocol.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Net-Minecraft-Login",
  "dist_version" => "0.001001",
  "license" => "perl",
  "module_name" => "Net::Minecraft::Login",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "HTTP::Tiny" => 0,
    "Moo" => 0,
    "Moo::Role" => 0,
    "Params::Validate" => 0,
    "Scalar::Util" => 0,
    "overload" => 0,
    "perl" => "v5.16.0",
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Test::Fatal" => 0,
    "Test::More" => "1.001002",
    "strict" => 0
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.4203",
  "Test::Fatal" => 0,
  "Test::More" => "1.001002",
  "strict" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Net/Minecraft/Role/HTTP.pm
use v5.16;
use warnings;

package Net::Minecraft::Role::HTTP {

  # ABSTRACT: Base class for Minecraft C<HTTP> things.

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Net::Minecraft::Role::HTTP",
    "interface":"role"
}

=end MetaPOD::JSON

=cut

  use Moo::Role;
  use HTTP::Tiny;
  use Scalar::Util qw( blessed );

=head1 CONSTRUCTOR ARGUMENTS

This section describes arguments that may be optionally passed to L<<< C<< ->new() >>|/new >>>, but as of the time of this writing, none are explicitly required,
and are offered only to give leverage to strange use cases ( and tests )

  my $instance = _SOME_CLASS_->new(
    user_agent   => ... ,
    http_headers => { ... },
    http_engine  => HTTP::Tiny->new(),
  );

=carg C<user_agent>

The User Agent to self-describe over HTTP

  type    : String
  default : "Net::Minecraft::Login/" . VERSION

=attr C<user_agent>

=cut

  has user_agent => (
    is      => rwp =>,
    lazy    => 1,
    default => sub {
      my $class = $_[0];
      $class = blessed($class) if blessed($class);
      my $version = $class->VERSION;
      $version = 'DEVEL' if not defined $version;
      return sprintf q{%s/%s}, $class, $version;
    },
  );

=carg C<http_headers>

Standard Headers that will be injected in each request

  type    : Hash[ string => string ]
  default : { 'Content-Type' => 'application/x-www-form-urlencoded' }

=attr C<http_headers>

=cut

  has http_headers => ( is => rwp =>, lazy => 1, default => sub { { 'Content-Type' => 'application/x-www-form-urlencoded' } }, );

=carg C<http_engine>

Low-Level HTTP Transfer Agent.

  type    : Object[ =~ HTTP::Tiny ]
  default : An HTTP::Tiny instance.

=attr C<http_engine>

=cut

  has http_engine => ( is => rwp =>, lazy => 1, default => sub { return HTTP::Tiny->new( agent => $_[0]->user_agent ) }, );

};

1;


lib/Net/Minecraft/LoginFailure.pm
use v5.16;
use warnings;

package Net::Minecraft::LoginFailure {

  # ABSTRACT: Result info for a Minecraft Login.

  use Moo;
  with 'Net::Minecraft::Role::LoginResult';

  use Carp qw( confess );
  use Params::Validate qw( validate SCALAR );
  use overload q{""} => 'as_string';


  sub is_success { return; }


  has code   => ( is => rwp =>, required => 1, isa => \&_defined_scalar_number );
  has reason => ( is => rwp =>, required => 1, isa => \&_defined_scalar );


  sub as_string {
    my ($self) = @_;
    return sprintf q[Login Failed: %s => %s], $self->code, $self->reason;
  }

  ## no critic ( RequireArgUnpacking RequireFinalReturn )
  sub _defined_scalar_number {
    confess q[parameter is not a defined value] unless defined $_[0];
    confess q[parameter is not a scalar] if ref $_[0];
    confess q[parameter is not a number] unless $_[0] =~ /^\d{1,3}$/;
  }

  sub _defined_scalar {
    confess q[parameter is not a defined value] unless defined $_[0];
    confess q[parameter is not a scalar] if ref $_[0];
  }
};
BEGIN {
  $Net::Minecraft::LoginFailure::AUTHORITY = 'cpan:KENTNL';
}
{
  $Net::Minecraft::LoginFailure::VERSION = '0.001001';
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Net::Minecraft::LoginFailure - Result info for a Minecraft Login.

=head1 VERSION

version 0.001001

=head1 CONSTRUCTOR ARGUMENTS

	my $error = Net::Minecraft::LoginFailure->new(
		code => $somecode,
		reason => $reason,
	);

This is ultimately a very low quality exception without throw on by default.

=head2 code

The HTTP Failure Code.

	type : HTTP Status Number ( ie: 000 -> 599 )

=head2 reason

The Reason given by the server for a Login Failure.

	type : String

=head1 METHODS

=head2 is_success

Always returns a false value for instances of this class.

=head2 as_string

	overload: for ""
	returns a string description of this login failure.

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Net::Minecraft::LoginFailure",
    "inherits":"Moo::Object",
    "does":"Net::Minecraft::Role::LoginResult",
    "interface":"class"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Net/Minecraft/Login.pm
use v5.16;
use warnings;

package Net::Minecraft::Login {

  # ABSTRACT: Basic implementation of the Minecraft Login Protocol.

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Net::Minecraft::Login",
    "inherits":"Moo::Object",
    "does":"Net::Minecraft::Role::HTTP",
    "interface":"class"
}

=end MetaPOD::JSON


=head1 DESCRIPTION

This is a basic implementation of the Minecraft Login protocol as described at L<http://wiki.vg/Session#Login>

  use Net::Minecraft::Login;

  my $ua = Net::Minecraft::Login->new();

  my $result = $ua->login(
    user => 'Bob',
    password => 'secret',
  );
  if( not $result->is_success ){
    die $result;
  }
  say "Login for user " . $result->user . " succeeded";

Note, it presently does no explicit session stuff, only performs the basic HTTP Request and returns the response as an object.


=cut

  use Moo;
  with 'Net::Minecraft::Role::HTTP';
  use Carp qw( confess );
  use Scalar::Util qw( blessed );
  use Params::Validate qw( validate SCALAR );
  use Net::Minecraft::LoginResult;
  use Net::Minecraft::LoginFailure;

=head1 CONSTRUCTOR ARGUMENTS

This section describes arguments that may be optionally passed to L<<< C<< ->new() >>|/new >>>, but as of the time of this writing, none are explicitly required,
and are offered only to give leverage to strange use cases ( and tests )

  my $instance = Net::Minecraft::Login->new(
    user_agent   => ... ,
    http_headers => { ... },
    http_engine  => HTTP::Tiny->new(),
    login_server => 'https://somewhere.else.org/'
    version      => 14, # IN THE FUTURE!
  );

Additional Constructor arguments can also be found in L<< C<Net::Minecraft::Role::HTTP>|Net::Minecraft::Role::HTTP >>

=carg C<login_server>

HTTP Address to authenticate with.

  type  : String
  default : https://login.minecraft.net/

=attr C<login_server>

=cut

  has login_server => ( is => rwp =>, lazy => 1, default => sub { return 'https://login.minecraft.net/' }, );

=carg C<version>

"Client" version.

  type  : String
  default : 13

This field indicates the version of the "Launcher". Minecraft may at some future time produce an updated launcher, and indicate that this specified version is out of date.

Mojang Minecraft Launchers will be required to download a newer version, and users of C<Net::Minecraft::Login> will either

=over 4

=item a) Be required to update to a newer C<Net::Minecraft::Login> that supports the newer version and changes that implies

=item b) Assuming no C<Login Protocol> Changes, only have to specify C<< version => >> either to the constructor, or as an argument to L</login>

=back

=attr C<version>

=cut

  ## no critic ( ProhibitMagicNumbers )
  has version => ( is => rwp =>, lazy => 1, default => sub { 13 }, );

=pmethod C<_do_request>

  signature : ( String $base_uri, Hash[ String => String ] $parameters , Hash[ String => Any ] $config )
  return    : Hash[ String => String ]

=cut

  sub _do_request {
    my ( $self, $base_uri, $parameters, $config ) = @_;
    my $uri = sprintf q{%s?%s}, $base_uri, $self->http_engine->www_form_urlencode($parameters);
    return $self->http_engine->request( GET => $uri, $config );

  }

=method C<login>

  signature: { user => String , password => String, version? => String }
  return   : Any( Net::Minecraft::LoginResult , Net::Minecraft::LoginFailure )

  my $result = $nmcl->login(
    user   => 'notch',
    password => 'jellybean',
  );

  if( not $result->is_success ){
    say "$result";
  } else {
    say "Logged in!";
  }

See L<< C<::LoginFailure>|Net::Minecraft::LoginFailure >> and L<< C<::LoginResult>|Net::Minecraft::LoginResult >>

=cut

  sub login {
    my ( $self, @args ) = @_;
    my %params = validate(
      @args => {
        user     => { type => SCALAR },
        password => { type => SCALAR },
        version  => { type => SCALAR, default => $self->version },
      }
    );
    my $result = $self->_do_request( $self->login_server, \%params, { headers => $self->http_headers }, );
    if ( not $result->{success} ) {
      my $failure = Net::Minecraft::LoginFailure->new(
        code   => $result->{status},
        reason => $result->{reason},
      );
      return $failure;
    }
    return Net::Minecraft::LoginResult->parse( $result->{content} );
  }
};
1;
